### ISSUE TITLE
Bug: Oqtane.Maui - A cross-origin resource sharing (CORS) was blocked of invalid or missing response headers.

### ISSUE THREAD

**Issue by thabaum** [oqtane/oqtane.framework#3636]
## Issue
Error while loading Oqtane.Maui Windows App:
```
Ensure CORS response header values are valid
A cross-origin resource sharing (CORS) request was blocked because of invalid or missing response headers of the request or the associated preflight request .
To fix this issue, ensure the response to the CORS request and/or the associated preflight request are not missing headers and use valid header values.
Note that if an opaque response is sufficient, the request's mode can be set to no-cors to fetch the resource with CORS disabled; that way CORS headers are not required but the response content is inaccessible (opaque).
```
## Screenshots
![image](https://github.com/oqtane/oqtane.framework/assets/13577556/d6fb0b30-ef39-4568-a49e-e3c7c954239e)

## How To Reproduce
1. Download current dev branch.
2. Open **2 Visual Studio instances** with Maui and Oqtane solutions loaded.
3. Set the `apiurl` to `http://localhost:44357` in the MauiConstants.cs file located in the Oqtane.Maui project.
4. **Build** both projects.
5. Fire up **Oqtane.Server**
6. Fire up **Oqtane.Maui**
7. **Hit F12** to review console logs and issues.

## Additional Context
I tried to work on this today to resolve debugging other behaviors.

I created a [branch relating to cors Oqtane.Maiu support](https://github.com/thabaum/oqtane.framework/tree/cors-maui-support)

I think this approach would work well.  If it sounds good I think it can be added per alias, or default alias, or maybe the alias needs to be CORS enabled and then gets added to the appsettings.json file.

Maybe a better solution, I tried services and such to just use adding the alias's in the database to this, however on startup maybe this be tough to do... as i found out.

**Comment by thabaum** [comment]
@sbwalker I created a branch I included a link to in the notes for this issue.  It is a work in progress so to speak as I think it would be a good discussion on how this should be implemented properly.

I just need to understand the behaviors we wish to include as I can create some logic to handle this in site settings.  "Enable CORS Support" option for default alias?  Just thinking out loud a little.  On initial site creation it can set this the same time Oqtane sets the default database with the first default alias so it is good to go out the box.  I just dont know if we need to add every alias, I think just the default should work correct?  Or that file could get massive...  We could also add another json file that only handles this.

I think it should be managed in the framework itself but not 100%. and I want to be sure before submitting any PR I can see if I am heading down the right path.  I can get this worked out tomorrow probably and move onto the file upload soon after.  I just like to refresh my code base from time to time and when working on this I want a repo I can load that works.  I would be happy to put this issue behind us if I can.

Hope this helps, cheers!

**Comment by sbwalker** [comment]
@thabaum in case I am  mistaken the only actual functionality issue related to CORS is the file upload scenario when using the MAUI client - there are no other functional issues which have been identified. It is not clear where the CORS issue identified above is originating but it does not seem to affect the run-time functionality.

In regards to how to fix the file upload issue I believe only the FileController Upload method should include an EnableCors attribute - there is no valid reason to enable CORS for the entire application as that would violate least privilege principles. The EnableCors attribute obviously needs to be associated to a policy which specifies the origins - and this is where it get complicated in a multi-tenant application. The policy origins need to be based on the Aliases which are defined for the Oqtane installation - which are stored in the master database. So during startup, the list of origins need to be constructed dynamically from the database. The next challenge is how to handle when a host user adds or removes an alias to a site - the app would need to be restarted because the CORS policy would need to be refreshed, which can only happen during startup in .NET Core. However restarts should never be triggered automatically as they can cause a lot of problems - so the host user would need to be notified that a restart is needed to finish the operation (much like the message when you install modules or themes). 

Anyways, those are the thoughts off the top of my head. A lot of complications for a single scenario which is only an issue in 1 specific hosting model (MAUI). Note that there is no need for any additional Site Settings or appsettings.json settings - the framework already has all of the information needed.

**Comment by sbwalker** [comment]
@thabaum I just realized that the MAUI client always runs as IP 0.0.0.0 - so the only origin which needs to be specified in the CORS policy is this specific IP address. So this should actually be very simple to implement as a CORS policy during startup - coupled with an EnableCors attribute on the File Upload method.

**Comment by thabaum** [comment]
> The policy origins need to be based on the Aliases which are defined for the Oqtane installation - which are stored in the master database. So during startup, the list of origins need to be constructed dynamically from the database. 

@sbwalker This was my first approach and failed trying to get things to load in an order that didnt make the framework cry.  I could get a solution to build but it would not run due to not being able to load alias's prior to running. 

> The policy origins need to be based on the Aliases which are defined for the Oqtane installation - which are stored in the master database. So during startup, the list of origins need to be constructed dynamically from the database. 

> I just realized that the MAUI client always runs as IP 0.0.0.0 - so the only origin which needs to be specified in the CORS policy is this specific IP address. So this should actually be very simple to implement as a CORS policy during startup - coupled with an EnableCors attribute on the File Upload method.

I was adding 0.0.0.0 as well but was concerned on the side of security allowing pretty much any app to be able to access this.  I was going to wait for this conversation to open up more.

This has no relation to the file uploads itself I don't think.  If anything maybe the fact the APP runs on HTTPS 0.0.0.0 and my site is running HTTP (Non SSL) which could be relating, but it does not resolve this error coming in dev tools relating to CORS.

I can share my code I created in another REPO with my approach on how I was trying to access Alias's and load them all as CORs Allowed Origins if you want to see what I was trying to do and maybe you can figure out the order of things to get it to work faster than I could or if it would even work.

So really I have made a few solutions and researched what I could relating to the topic which aligns with your thoughts entirely.

**Comment by sbwalker** [comment]
@thabaum I think I explained most of this previously in various threads...

1. MAUI client runs as IP 0.0.0.0 - it tries to connect to server running at localhost:44357 (or some other address) using XMLHttpRequest (JavaScript) - this results in a CORS issue as the origin does not match the server
2. Oqtane Server needs to be configured to allow CORS requests from the origin - which is 0.0.0.0 (no other origins should be allowed or specified)
3. Only those specific API methods requiring CORS should be exposed using the [EnabledCors] attribute - ie. only those API methods which need to service XMLHttpRequest (JavaScript) requests from the MAUI client. There will be very few methods which require this, as the majority of API calls are done via HttpClient (.NET)
4. No additional Site Settings or config options should be created

**Comment by thabaum** [comment]
@sbwalker I can understand this approach fully the thing is:
1. is 0.0.0.0 is a placeholder for any origin same as `builder.AllowAnyOrigin();` is `builder.WithOrigins(0.0.0.0);`  maybe this should only be enabled if someone wants to open things up relating to CORS?  This is ok for development, I am not sure OK with production, although the API will have defensive logic it sounds to help, it is really almost useless.  Entering the code below in startup.cs removes the error.
2. If you put the alias of the site in place of 0.0.0.0 it works too. No CORS error and maybe best practices for production... which is why I made the release file blank to be filled in as alias's are created.
3. I am not sure how many apps will run per instance of oqtane, per alias... which is why I was suggesting an "enable" feature possibly to allow this for the site, and let them know there is a restart, and add the alias to the json config, instead of all sites ever created on an instance.  Remote App Ready? type of setting.

Like you said 0.0.0.0 allows them all.  I just want to double check this is going to be the solution that wont bring up other issues.

Thanks for the extra explanation as it clears up some steps on connectivity.

You are suggesting `[EnableCors("CorsPolicySettings")] ` be included on the api controllers where this type of activity needs checked the most.   I will test this with file uploads as you seem to think it maybe related.

**Comment by sbwalker** [comment]
@thabaum in the case of .NET MAUI, as far as I know it ALWAYS runs as IP 0.0.0.0 (there is no way to change this). Since the only external client app which Oqtane supports is .NET MAUI then the only origin which needs to be supported is 0.0.0.0. There is no need to specify alias values as those are SERVER domains - not originating client domains. And I am saying that EnableCors should be added at a METHOD level - not CONTROLLER (Controller is too general).

**Comment by thabaum** [comment]
@sbwalker I just want to point out, and I understand where you are coming from here that using the ALIAS resolves the issue for the MAUI app.  If I create an APP using MAUI, the my app will pass that Oqtane CORS policy you are suggesting.  However, if you set the domain name the MAUI app actually uses this domain as the originating IP as well.   Sometimes I find 0.0.0.0 and other times I get localhost address for where the originating url came from if I recall from studying a few test scenarios I created yesterday and reviewing errors.

So in other words, this is why using the app alias works and why I put on the brakes to 0.0.0.0

**Comment by sbwalker** [comment]
@thabaum I don't think you fully understand the interactions between client and server and the differences in run-time behavior when razor components are executed on the MAUI client vs a web browser. 

1. When you set the domain name for the MAUI client app on the local device in the appsettings.json, it sets the HttpClient BaseAddress during startup. HttpClient is used for API calls originating in C# code on the MAUI client app. Because the origin (ie. BaseAddress) on the client matches the domain on the server, there is no CORS issues

2. Requests for static assets (ie. images, CSS, JS) are all converted into absolute Urls when running the MAUI client app (because if they were not absolute Urls they would be treated as relative Urls to assets on the client device which would result in 404 errors).  Static assets are handled by UseStaticFiles and do not have CORS issues.

3. These above 2 types of requests cover the majority of interactions between the client and server. However there is one other type of request in Oqtane - XMLHttpRequest calls originating in JavaScript code which is called by JS Interop. In this specific case the origin is IP 0.0.0.0 and it is calling an API on the server which is running under a different domain. This triggers a CORS issue. The only example of this so far which has been identified is the file upload method.

So 3. is the only scenario where CORS needs to be handled. And since the origin is always 0.0.0.0 when running .NET MAUI, it is the only origin which needs to be allowed by the CORS policy on the server.

**Comment by thabaum** [comment]
@sbwalker thanks for the clarification, there was some confusion and I think I have a further understanding of the types.

I took a stab at seeing if this would resolve file uploads and no luck there.  I learned my order was wrong once I applied this to the controller itself.

The profile for HOST for example would not load until I put it globally more or less so looking at which need it will take some time.

I tried to just do all the post methods but I could not stay authenticated with authorization in anything relating to looking at files.

**Comment by sbwalker** [comment]
When running .NET MAUI you can use F12 to launch the Edge Dev Tools to see what is happening in the embedded WebView. I captured the following information during the file upload in MAUI:

![image](https://github.com/oqtane/oqtane.framework/assets/4840590/0d11be61-3f2b-413b-b89f-4ac298e9d713)


![image](https://github.com/oqtane/oqtane.framework/assets/4840590/2b115b07-5cf3-4d6c-ba29-304c701063f6)

**Comment by thabaum** [comment]
@sbwalker Yes I did see these errors and submitted a few PRs relating just now for you to review.

I had the CORS issue not a problem and still had the null which I think is the POST error at the bottom relating...

Both `folder` and `formfile` are null after getting past the CORS error.

**Comment by thabaum** [comment]
I got my IDE to use SSL and fire up oqtane using https://localhost:44357

Now the fun stuff:
```
18:18:50:532	Microsoft.AspNetCore.Cors.Infrastructure.CorsService: Information: CORS policy execution failed.
18:18:50:532	Microsoft.AspNetCore.Cors.Infrastructure.CorsService: Information: Request origin https://localhost:44357 does not have permission to access the resource.
```

This is on the Oqtane.Server VS instance.

not sure how that error got there but it is during my SSL testing.  I have VS Oqtane.Server running HTTPS and the Maui app trying to connect to it.

```
18:42:12:241	Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Net.Security.dll
18:42:12:241	Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Private.CoreLib.dll
18:42:12:241	Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Private.CoreLib.dll
18:42:12:241	Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Private.CoreLib.dll
18:42:12:502	Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Net.Security.dll
18:42:12:502	Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Private.CoreLib.dll
18:42:12:502	Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Net.Security.dll
18:42:12:502	Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Private.CoreLib.dll
18:42:12:502	Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Private.CoreLib.dll
18:42:12:502	Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Private.CoreLib.dll
18:42:12:502	Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Net.Security.dll
18:42:12:502	Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Private.CoreLib.dll
18:42:12:502	Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Net.Security.dll
18:42:12:502	Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Private.CoreLib.dll
18:42:12:502	Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Private.CoreLib.dll
18:42:12:502	Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Private.CoreLib.dll
18:42:12:502	Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Net.Security.dll
18:42:12:502	Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Private.CoreLib.dll
18:42:12:754	Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Net.Security.dll
18:42:12:754	Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Private.CoreLib.dll
18:42:12:754	Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Private.CoreLib.dll
18:42:12:754	Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Private.CoreLib.dll
18:42:12:754	Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Net.Security.dll
18:42:12:754	Exception thrown: 'System.Security.Authentication.AuthenticationException' in System.Private.CoreLib.dll
```

These error happen when I hit "Reload" after clearing the debug info as shown below:
![image](https://github.com/oqtane/oqtane.framework/assets/13577556/41cd2025-d0aa-45dd-b784-7d104ae56f8e)

This is just trying to review things that may relate, and really SSL should be default enabled on our environments.  I just had to edit the debug configuration.

I updated the ApiUrl in MauiConstants.cs to use HTTPS as well, it is trying to connect but cannot.

I wish I could reproduce the first screenshot, I had to show that to let you know this happened.  I think what I did to resolve it is I was using origin localhost:* instead of localhost:44357 which may have got that error removed.  I was trying to see if adding the alias would make a difference I believe is how I produced it.

I will be setting up a production test site tomorrow.

**Comment by thabaum** [comment]
I have been testing some things maybe found another odd issue with port 80 and port 443 requests looping I believe I meantioned in another issue.   If I am not logged into the site, the site is fine, but if I am logged in and leave the page for a while it will get unresponsive and even has crashed out the site once causing it to restart.  Not 100% sure just felt like that may have happened.  I will get more details but thought it maybe a clue to investigate further.

**Comment by thabaum** [comment]
@sbwalker I was able to resolve this so it is not an issue, but relating to AntiForgery I was wondering if             `app.UseAntiforgery();` should be included in the **Startup.cs** after `app.UseAuthorization()` or is it not necessary?

**Comment by sbwalker** [comment]
@thabaum I am having a very difficult time following the information in this thread. The original post appears to be a CORS issue that is related to loading the initial page within .NET MAUI. However there is no indication of what functionality is actually broken. And it is not clear from the other posts in this thread what is the root cause. ie. is it related to the client app using a scheme which is different from the server (HTTPS -> HTTP). Or is it related to a missing HTTP header. Or something else?

**Comment by thabaum** [comment]
The functionality that is broken: no communication with the file controller allowed to upload files.

**Comment by sbwalker** [comment]
@thabaum the title of this issue and original steps to reproduce have nothing to do with file uploading - it appears to be related to CORS and the initial page load of the .NET MAUI app. 

It is important that issue reports are very specific and focused and do not make assumptions or include a wide variety of unrelated behaviors or scenarios.

I created #3669 for file uploading on .NET MAUI. I also documented my specific observations on why the issue is occurring. I believe I will have a working solution tomorrow.
